Fix Linux ARM64 build

ivankra opened the pull request:
It builds with clang 18-22 on recent Debian/Ubuntu, both with and without ICU, passes tests and benchmarks.

Key issues:

* Different ARM64 vararg calling convention between macOS (DarwinPCS) and Linux (AAPCS64).

  Fixed CALL_ENTRYPOINT_NOASSERT and DECLARE_ARGS_VARARRAY using va_list.__stack from the official ABI - robust and compiler-independent.

  However, there is also JavascriptStackWalker.cpp which depends on the exact stack layout and magic constants, especially ArgOffsetFromFramePtr, this part is more fragile.

* char is unsigned in Linux ARM64 ABI unlike macOS/Win, breaking int8 code. Make sure __int8 (=char) is always prefixed with signed/unsigned.

* arm64/\*.S files use Microsoft-style ; comments, unsupported by GNU assembler. These were already converted in amd64/\*.S files to //, but I propose a simpler fix to just strip them on the fly during the build with sed.

* Missing _GetNativeSigSimdContext based on https://github.com/dotnet/runtime/blob/main/src/coreclr/pal/src/thread/context.cpp#L927

* Cpsr register is PState on Linux

* wchar_t/char16_t mismatches when building with ICU - caused by libstdc++ on Debian undefining macros like wcslen. Solved by including problematic system headers early in pal.h.

* Minor compile errors and bugs fixed along the way: MAX_INT is not an exact float, UNumberFormatFields cast, strict aliasing bug in LuHiDbl/LuLoDbl, TZ=US/Pacific doesn't always exist, __reserved in sal.h conflicts with signal.h.

Note: binary with JIT crashes - this must be built with ./build.sh --no-jit. cmake already prints a warning that ARM64 JIT is only supported on Windows.

Tested as follows:

```
cat >Dockerfile <<EOF
ARG BASE=ubuntu:24.04
FROM $BASE
RUN export DEBIAN_FRONTEND=noninteractive && apt-get update -y && apt-get install -y --no-install-recommends build-essential ca-certificates clang cmake git libicu-dev llvm ninja-build python3 tzdata
EOF

BASE=ubuntu:24.04; podman build --build-arg BASE=$BASE -t chakra-$BASE . && podman run --rm -it -v $PWD:$PWD -w $PWD chakra-$BASE /bin/bash -c "rm -rf out; ./build.sh --ninja --no-jit && ninja -C out/* check"
```

For these combinations: ubuntu:22.04 (clang 14), ubuntu:24.04 (clang 18), ubuntu:26.04 (clang 21), debian:stable (clang 19), debian:sid (clang 21), and three build commands: ./build.sh --ninja --no-jit, ./build.sh --ninja --no-jit --debug, ./build.sh --ninja --no-jit -t.

ubuntu:22.04 with ./build.sh --ninja --no-jit -t crashes in JavascriptStackWalker, but everything else passes.

Benchmarking on Octane test suite is on par with JIT-less SpiderMonkey (on Mac M4):

| Benchmark    | chakracore_jitless | sm_jitless | v8_jitless | quickjs  |
|--------------|--------------------|------------|------------|----------|
| Richards     | 566.27             | 508.91     | 2050.18    | 1748.91  |
| DeltaBlue    | 509.45             | 524.45     | 2224.36    | 1594.91  |
| Crypto       | 895.73             | 809.82     | 1902.64    | 2277.55  |
| RayTrace     | 1360.36            | 1717.91    | 7105       | 3132.91  |
| EarleyBoyer  | 2055.25            | 2389       | 11803.75   | 3840     |
| RegExp       | 2533.75            | 1012.25    | 5484.75    | 859.50   |
| Splay        | 4166               | 3533.25    | 12368.50   | 4153.50  |
| NavierStokes | 1240.50            | 1484.75    | 3473.25    | 4385.25  |
| gmean        | 1324.31            | 1213.76    | 4510.95    | 2425.15  |

Fixes #7050

ivankra committed:
Fix Linux ARM64 build

Key issues during port:

* Different ARM64 vararg calling convention between macOS (DarwinPCS)
  and Linux (AAPCS64).

  Fixed CALL_ENTRYPOINT_NOASSERT and DECLARE_ARGS_VARARRAY using
  va_list.__stack from the official ABI - robust and compiler-independent.

  However, there is also JavascriptStackWalker.cpp which depends on the
  exact stack layout and magic constants, especially ArgOffsetFromFramePtr,
  this part is more fragile.

* char is unsigned in Linux ARM64 ABI unlike macOS/Win, breaking int8 code.
  Make sure __int8 (=char) is always prefixed with signed/unsigned.

* arm64/*.S files use Microsoft-style ; comments, unsupported by GNU
  assembler. These were already converted in amd64/*.S files to //,
  but I propose a simpler fix to just strip them on the fly during
  the build with sed.

* Missing _GetNativeSigSimdContext based on
  https://github.com/dotnet/runtime/blob/main/src/coreclr/pal/src/thread/context.cpp#L927

* Cpsr register is PState on Linux

* wchar_t/char16_t mismatches when building with ICU.
  For now solved with #define wcslen PAL_wcslen etc in ChakraICU.h.
  It builds at least, though some Intl tests are failing.

Note: binary with JIT crashes - this must be built with ./build.sh --no-jit.
cmake already prints a warning that ARM64 JIT is only supported on Windows.

rhuanjl commented on lib/Runtime/PlatformAgnostic/ChakraICU.h:
```diff
@@ -46,6 +46,20 @@
 #include "unicode/upluralrules.h"
 #endif // ifdef WINDOWS10_ICU
 
+// Use PAL wrappers for Linux arm64 to fix wchar_t/char16_t mismatches.
```
This shouldn't be needed as pal.h should be doing these defines; needing to add them here suggests the include order is wrong, additionally I'd be surprised if there is anything arm specific about the include order.

Could there be an include order issue with a newer version of clang (vs what we use in CI) or with a different flavour of linux (CI only tests Ubuntu)?

ivankra replied:
From my testing, it looks like this needs to go anywhere inside lib/Runtime/Runtime.h between PlatformAgnostic/ChakraPlatform.h and Library/JavascriptFunction.h.

Putting it just before before ChakraPlatform.h (vs immediately after) breaks the build in a spooky action at a distance way:
```
In file included from /home/ivan/src/ChakraCore/lib/Jsrt/Core/JsrtCore.cpp:20:
/home/ivan/src/ChakraCore/lib/Common/Codex/Utf8Helper.h:223:27: error: no matching function for call to 'wcslen'
  223 |       [  7%] Building CXX object lib/Common/DataStructures/CMakeFiles/Chakra.Common.DataStructures.dir/DictionaryStats.cpp.o
      sourceString, wcslen(sourceString), destStringPtr, &unused);
      |                           ^~~~~~
/usr/include/wchar.h:247:15: note: candidate function not viable: no known conversion from 'LPCWSTR' (aka 'const char16_t *') to 'const wchar_t *' for 1st argument
  247 | extern size_t wcslen (const wchar_t *__s) __THROW __attribute_pure__;
      |               ^       ~~~~~~~~~~~~~~~~~~
/home/ivan/src/ChakraCore/lib/Jsrt/Core/JsrtCore.cpp:919:18: error: no matching function for call to 'wcslen'
  919 |         length = wcslen((const char16 *)content);
      |                  ^~~~~~
/usr/include/wchar.h:247:15: note: candidate function not viable: no known conversion from 'const char16 *' (aka 'const char16_t *') to 'const wchar_t *' for 1st argument
  247 | extern size_t wcslen (const wchar_t *__s) __THROW __attribute_pure__;
      |               ^       ~~~~~~~~~~~~~~~~~~
```

This is probably due to various `#pragma once`, Pch.h system and headers being generously over-included all over the place. There's *probably* a right place somewhere deep in pal where this can be properly fixed, but it is hard to identify it when the build behaves in this way.

ChakraPlatform.h also transitively includes ChakraICU.h and redefining wcs* functions before unicode headers  breaks the build in STL headers (perhaps libstdc++ issue? macOS/Windows build I suppose default to clang's libc++):
```
In file included from /home/ivan/src/ChakraCore/lib/Runtime/PlatformAgnostic/Platform/Common/UnicodeText.Common.cpp:8:
In file included from /home/ivan/src/ChakraCore/lib/Runtime/PlatformAgnostic/UnicodeText.h:8:
In file included from /home/ivan/src/ChakraCore/lib/Runtime/PlatformAgnostic/ChakraICU.h:52:
In file included from /usr/include/unicode/ucol.h:1525:
In file included from /usr/lib/gcc/aarch64-linux-gnu/14/../../../../include/c++/14/string_view:48:
In file included from /usr/lib/gcc/aarch64-linux-gnu/14/../../../../include/c++/14/bits/char_traits.h:42:
In file included from /usr/lib/gcc/aarch64-linux-gnu/14/../../../../include/c++/14/bits/postypes.h:40:
In file included from /usr/lib/gcc/aarch64-linux-gnu/14/../../../../include/c++/14/cwchar:44:
/usr/include/wchar.h:130:12: error: exception specification in declaration does not match previous declaration
  130 | extern int wcscmp (const wchar_t *__s1, const wchar_t *__s2)
      |            ^
/home/ivan/src/ChakraCore/lib/Runtime/PlatformAgnostic/ChakraICU.h:14:17: note: expanded from macro 'wcscmp'
   14 | #define wcscmp  PAL_wcscmp
      |                 ^
/home/ivan/src/ChakraCore/pal/inc/pal.h:6242:23: note: previous declaration is here
 6242 | PALIMPORT int __cdecl PAL_wcscmp(const WCHAR*, const WCHAR*);
      |                       ^
```

Library/JavascriptFunction.h included in Runtime.h needs these wcs* defines already, so they need to be in place before that point:
```
In file included from /home/ivan/src/ChakraCore/lib/Runtime/Library/JavascriptFunction.cpp:5:
In file included from /home/ivan/src/ChakraCore/lib/Runtime/Library/RuntimeLibraryPch.h:10:
In file included from /home/ivan/src/ChakraCore/lib/Runtime/./Runtime.h:436:
/home/ivan/src/ChakraCore/lib/Runtime/./Library/JavascriptFunction.h:288:12: error: no matching function for call to 'wcscmp'
  288 |         if(wcscmp(displayName, Js::Constants::AnonymousFunction) == 0)
      |            ^~~~~~
/usr/include/wchar.h:130:12: note: candidate function not viable: no known conversion from 'PCWSTR' (aka 'const char16_t *') to 'const wchar_t *' for 1st argument
  130 | extern int wcscmp (const wchar_t *__s1, const wchar_t *__s2)
      |            ^       ~~~~~~~~~~~~~~~~~~~
1 error generated.
```

ivankra replied:
Oh, I found the culprit in /usr/include/c++/14/cwchar:
```
$ rg def.*wcs /usr/include/
/usr/include/wchar.h
546:#  define wcstol __isoc23_wcstol
547:#  define wcstoul __isoc23_wcstoul
...
/usr/include/c++/14/cwchar
// Get rid of those macros defined in <wchar.h> in lieu of real functions.
...
81:#undef mbsrtowcs
100:#undef wcscat
101:#undef wcschr
102:#undef wcscmp
103:#undef wcscoll
...
```

So, looks like an issue caused by libstd++ in recent Debian versions.

Tested build on Ubuntu 22.04 - it works without these extra defines in ChakraICU.h.

rhuanjl commented on CMakeLists.txt:
```diff
@@ -522,9 +532,6 @@ else()
 endif()
 
 if(CC_TARGETS_ARM64)
-    if(CC_TARGET_OS_LINUX)
```
I'd like to see CI running before we remove this warning; I think we should be able to add linux on arm to the cirrus CI.

ivankra replied:
Sure, I will bring it back.

rhuanjl commented on lib/Runtime/Language/Arguments.h:
```diff
@@ -84,8 +92,19 @@ inline int _count_args(const T1&, const T2&, const T3&, const T4&, const T5&, Js
+// Linux ARM64 uses AAPCS64: first 8 args in x0-x7, rest via stack.
+// Fill x2-x7 with nulls here to force the expected stack layout:
+// [RetAddr] [function] [callInfo] [args...]
+#define CALL_ENTRYPOINT_NOASSERT(entryPoint, function, callInfo, ...) \
```
I'd expect this to break the JavascriptStackWalker, I saw you said it was "fragile", what's the status?

ivankra replied:
The single hard-coded constant ArgOffsetFromFramePtr that JavascriptStackWalker relies on feels rather fragile to me. For now, I set ArgOffsetFromFramePtr=4 - I see 8 extra bytes of what seems like padding or saved registers. But I'd expect it to break with different compilers and optimization flags by analogy with my experience with DECLARE_ARGS_VARARRAY - at first (before I found va_list.__stack) I was trying to fixing it by trying different offsets in _get_va(), but it just kept breaking between optimized and debug builds.

JavascriptStackWalker in JIT-less builds so far works for me with clang 19, 20, and 22, both optimized and debug builds. Perhaps because it only needs to deal with a few specific call sites in this mode. When JIT is enabled it is crashing, but because JIT wasn't supported on macOS, I didn't investigate these crashes further.

rhuanjl replied:
The JIT crash is not remarkable, the ARM JIT contains significant windows specific logic that needs re-writing for Linux/macOS including the way function entry and exit is setup and the managing of the stack layout.

For calling JS functions in the interpreter we go via arm64_CallFunction to set the stack exactly the way we need for the stack walker and argument handling to work: https://github.com/chakra-core/ChakraCore/blob/master/lib/Runtime/Library/arm64/arm64_CallFunction.S

rhuanjl commented on test/Number/toString.js:
```diff
@@ -23,7 +23,11 @@ function runTest(numberToTestAsString)
     writeLine("n.toString(2):  " + n.toString(2));
     writeLine("n.toString(16):  " + n.toString(16));
-    writeLine("n.toString(25):  " + n.toString(25));
+    if (!numberToTestAsString.endsWith('e21')) {
```
I hit this same issue on macOS ARM, and disabled the test for mac see  test/Number/rlexe.xml lines 17-24.

Awkwardly we don't have a mechanism for disabling a test for an OS and CPU combination.

ivankra replied:
Ok, I'll try reenabling that test for macOS then with this change.

rhuanjl reviewed (commented):
Thank you for your work on this; it's certainly good to support another platform.

I've put a few initial comments above though I'm going to want to another slightly more detailed review.

Are you able to sign our Contributor agreement? https://github.com/chakra-core/ChakraCore/blob/master/ContributionAgreement.md

ivankra committed:
CLA, copyrights, ARM64 linux warning, reenable toString.js on macOS.

ivankra commented:
> Are you able to sign our Contributor agreement? https://github.com/chakra-core/ChakraCore/blob/master/ContributionAgreement.md

Yes, added my name there.

ivankra committed:
Fix PAL_DispatchException linker error

Seemingly only needed for macOS. Linker complains about references to it
under some configurations.

ivankra committed:
Fix UnsetField compile error under newer clang

ivankra committed:
Alternative fix for unicode errors

Include system headers that may cause to PAL_* macros to
get undefined early in pal.h.

ivankra committed:
Fix DateTimeFormat.js error: TZ=US/Pacific -> America/Los_Angeles

US/Pacific is a legacy alias, unavailable in the standard
installation of Debian 13, which moved it to tzdata-legacy.

ivankra committed:
Fix ParseInt1.js error

parseInt("4294967296") returned 1 due to strict aliasing bug at BigUInt.cpp:529.

Fix LuHiDbl/LuLoDbl to prevent similar issues caused by them:
access uint32 via a union, and __may_alias__ annotation for clang.

ivankra commented:
I fixed the remaining errors that I encountered. 

For the unicode errors: they were ultimately caused by Debian's libstdc++ undefining wcslen etc macros. I changed the fix to include the problematic system headers early in pal.h: https://github.com/ivankra/ChakraCore/commit/6bba7eea81c4d981767e942ba69c71a9dd6d6b96

Tested commit [c8cce2106209880d1c8b8d9f3f805c9c01eaeb37](https://github.com/chakra-core/ChakraCore/commit/c8cce2106209880d1c8b8d9f3f805c9c01eaeb37) in three build modes: `./build.sh --ninja --no-jit`, `./build.sh --ninja --no-jit --debug`, `./build.sh --ninja --no-jit -t` with command like this:

```sh
cat >Dockerfile <<EOF
ARG BASE=ubuntu:24.04
FROM $BASE
RUN export DEBIAN_FRONTEND=noninteractive && apt-get update -y && apt-get install -y --no-install-recommends build-essential ca-certificates clang cmake git libicu-dev llvm ninja-build python3 tzdata
EOF

BASE=ubuntu:24.04; podman build --build-arg BASE=$BASE -t chakra-$BASE . && podman run --rm -it -v $PWD:$PWD -w $PWD chakra-$BASE /bin/bash -c "rm -rf out; ./build.sh --ninja --no-jit && ninja -C out/* check"
```

On these userlands: ubuntu:22.04 (clang 14), ubuntu:24.04 (clang 18), ubuntu:26.04 (clang 21), debian:stable (clang 19), debian:sid (clang 21).

Everything builds and passes all disable_jit tests, except for a single combination: ubuntu:22.04 with `./build.sh --ninja --no-jit -t` crashes in JavascriptStackWalker. Looks like stack layout mismatch, and more profound than a merely wrong ArgOffsetFromFramePtr. I'd suggest to ignore it since it uses a rather old clang, and on newer systems everything passes.

rhuanjl commented:
@ppenzin any thoughts?

I'd like to do one more read through later before merge to check for any oddities I've missed.

But unless I notice anything else, I think we should bring this in and then maybe one of us could open a new PR to add CI for it (and remove the build time warning assuming we get CI to pass).

rhuanjl commented:
> Everything builds and passes all disable_jit tests, except for a single combination: ubuntu:22.04 with `./build.sh --ninja --no-jit -t` crashes in JavascriptStackWalker. Looks like stack layout mismatch, and more profound than a merely wrong ArgOffsetFromFramePtr. I'd suggest to ignore it since it uses a rather old clang, and on newer systems everything passes.

I'm always a bit nervous when there's an error we can't understand - if the stack walker is not working it's likely an incompatibility between the Call entry point handlers in Arguments.h and arm64_CallFunction; but it's odd that it would work on the newer compilers but not Clang14.

ppenzin commented:
Ubuntu 22.04 is going to go into security-only mode relatively soon, that can make reproducing this a bit difficult in the future, that's why I'm not sure how important it would be to fix. @ivankra how much of test suite is failing?

This PR fixes Clang 21 miscompilation that we recently found. Overall it looks reasonable, but I will take another look at it.

ivankra commented:
Much of the test suite fails in that particular combination:
```
$ export BASE=ubuntu:22.04; podman build --build-arg BASE=$BASE -t chakra-$BASE . && podman run --rm -it -v $PWD:$PWD -w $PWD chakra-$BASE /bin/bash -c "rm -rf out; ./build.sh --ninja --no-jit -t && ninja -C out/* check" 
...
############ Results for disable_jit tests ###########
...
# Total: passed 1120, failed 835
```

So, in a way, it would be be easy for a user to notice something is wrong with a built binary.

rhuanjl reviewed (approved):
If/when we remove the existing CMake warning we can put in a Clang version check and warn for versions less than 15 or something.

Waiting on @ppenzin's review before we merge.

rhuanjl commented on pal/src/thread/context.cpp:
```diff
@@ -545,6 +554,47 @@ CONTEXT_SetThreadContext(
 
+#if defined(__linux__) && defined(_ARM64_)
+// Reference: https://github.com/dotnet/runtime/blob/main/src/coreclr/pal/src/thread/context.cpp#L927
+static inline fpsimd_context* _GetNativeSigSimdContext(unsigned char* data, size_t size)
+{
+    size_t pos = 0;
+    while (pos < size)
+    {
+        _aarch64_ctx* ctx = reinterpret_cast<_aarch64_ctx*>(&data[pos]);
+        if (pos + sizeof(_aarch64_ctx) > size)
+        {
+            break;
+        }
```
I had another look at this, and I was wondering why it's so different to the function it's based on in dotnet, does the dotnet version require something we don't have or did it contain a performance trap?

ivankra replied:
I rewrote it from scratch. Dotnet's version is more general returning  fpsimd_context and sve_context, but Chakra only needs fpsimd_context.

ppenzin reviewed (approved):
LGTM

ppenzin commented:
@rhuanjl, not sure I understand your comment about CMake warning, aside from that I don't have any issues with merging this PR.

rhuanjl committed:
Merge branch 'master' into linux-arm64
